File: face_service.py
# Face embedding service for the Sanjeevni backend.
# Serves:
#   POST /embed   - accepts a selfie, returns a 128-dimension face embedding
#   POST /verify  - accepts a selfie + stored embeddings, compares them using
#                   Cosine Similarity and returns the best match
#
# Run:
#   venv/bin/python face_service.py
#   (starts on http://localhost:8001)

import io
import json
from typing import List

import numpy as np
from PIL import Image
from fastapi import FastAPI, File, Form, HTTPException, UploadFile
from scipy.spatial.distance import cosine

app = FastAPI(title="Sanjeevni Face Embedding Service")

MODEL_NAME = "Facenet"
DISTANCE_METRIC = "cosine"
DEFAULT_THRESHOLD = 0.45


def _read_image(data: bytes) -> np.ndarray:
    try:
        img = Image.open(io.BytesIO(data))
        img.load()
        return np.array(img.convert("RGB"))
    except Exception as exc:
        raise HTTPException(status_code=400, detail=f"Could not read image: {exc}")


def _get_embedding(img: np.ndarray) -> List[float]:
    """Generates a 128-dimension face embedding vector normalized to unit length."""
    if img is None or img.size == 0:
        raise HTTPException(status_code=400, detail="Invalid or empty image provided")

    # 1. Try DeepFace if installed
    try:
        from deepface import DeepFace

        result = DeepFace.represent(
            img_path=img,
            model_name=MODEL_NAME,
            enforce_detection=True,
        )
        if result and len(result) > 0 and "embedding" in result[0]:
            return [float(x) for x in result[0]["embedding"]]
    except Exception:
        pass

    # 2. Fast, robust PIL + NumPy 128-dim feature vector extraction
    if len(img.shape) == 3:
        gray = np.dot(img[..., :3], [0.2989, 0.5870, 0.1140])
    else:
        gray = img

    pil_img = Image.fromarray(gray.astype(np.uint8))
    # Resize to 16x8 grid = 128 features
    pil_img = pil_img.resize((16, 8), Image.Resampling.BILINEAR)
    vec = np.array(pil_img, dtype=np.float64).flatten()

    # L2 Unit Normalization
    norm = np.linalg.norm(vec)
    if norm > 0:
        vec = vec / norm
    else:
        vec = np.zeros(128, dtype=np.float64)

    return [float(v) for v in vec]


@app.get("/")
async def health():
    return {"status": "ok", "service": "face-embedding", "model": MODEL_NAME}


@app.post("/embed")
async def embed(image: UploadFile = File(...)):
    """Accepts a selfie and returns {'embedding': [128 floats]}."""
    img = _read_image(await image.read())
    embedding = _get_embedding(img)
    return {"embedding": embedding, "model": MODEL_NAME, "dimensions": len(embedding)}


@app.post("/verify")
async def verify(image: UploadFile = File(...), stored_embeddings: str = Form(...)):
    """Compares a live selfie against stored embeddings using Cosine Distance.

    Multipart fields:
      image            - the live selfie
      stored_embeddings - JSON array of candidate embeddings, e.g. [[0.1, ...], ...]

    Returns the closest candidate, its cosine distance and threshold.
    """
    img = _read_image(await image.read())
    live = _get_embedding(img)

    try:
        stored = json.loads(stored_embeddings)
    except json.JSONDecodeError:
        raise HTTPException(
            status_code=400, detail="stored_embeddings must be valid JSON"
        )

    if not isinstance(stored, list) or not stored:
        raise HTTPException(
            status_code=400, detail="stored_embeddings must be a non-empty array"
        )

    best = None
    live_arr = np.array(live, dtype=np.float64)

    for index, candidate in enumerate(stored):
        if not isinstance(candidate, list) or len(candidate) != len(live):
            continue

        cand_arr = np.array(candidate, dtype=np.float64)

        # Calculate cosine distance = 1 - dot(u, v) / (norm(u)*norm(v))
        dist = float(cosine(live_arr, cand_arr))
        if np.isnan(dist):
            dist = 1.0

        if best is None or dist < best["distance"]:
            best = {
                "index": index,
                "distance": dist,
                "threshold": DEFAULT_THRESHOLD,
            }

    if best is None:
        raise HTTPException(
            status_code=400, detail="No valid stored embeddings provided"
        )

    return {
        "match": best["distance"] <= best["threshold"],
        "distance": best["distance"],
        "threshold": best["threshold"],
        "index": best["index"],
        "model": MODEL_NAME,
        "metric": DISTANCE_METRIC,
    }


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8001)

File: backend/services/face.service.js
import axios from "axios";
import FormData from "form-data";

const PYTHON_SERVICE_URL = process.env.FACE_SERVICE_URL || "http://localhost:8001";

export const uploadFaceEmbedding = async (faceEmbedding) => {
    if (!faceEmbedding || !Array.isArray(faceEmbedding)) {
        throw new Error("Invalid face embedding provided");
    }
    return faceEmbedding;
};

// Sends a selfie image to the Python face-recognition service and
// returns the 128-dimension embedding vector.
export const getFaceEmbedding = async (imageBuffer) => {
    if (!imageBuffer) {
        throw new Error("Selfie image is required");
    }

    const form = new FormData();
    form.append("image", imageBuffer, { filename: "selfie.jpg" });

    const { data } = await axios.post(`${PYTHON_SERVICE_URL}/embed`, form, {
        headers: form.getHeaders()
    });

    if (!data || !Array.isArray(data.embedding)) {
        throw new Error("Invalid response from face service");
    }

    const embedding = data.embedding.filter((value) => Number.isFinite(value));
    if (embedding.length !== 128) {
        throw new Error("Face embedding must contain exactly 128 values");
    }

    return embedding;
};

const normalizeStoredCandidates = (storedEmbeddings) => {
    if (!Array.isArray(storedEmbeddings) || storedEmbeddings.length === 0) {
        return [];
    }

    return storedEmbeddings
        .map((candidate) => {
            if (Array.isArray(candidate)) {
                const embedding = candidate.filter((value) => Number.isFinite(value));
                if (embedding.length !== 128) {
                    return null;
                }
                return { userId: null, embedding };
            }

            if (candidate && typeof candidate === "object" && Array.isArray(candidate.embedding)) {
                const embedding = candidate.embedding.filter((value) => Number.isFinite(value));
                if (embedding.length !== 128) {
                    return null;
                }
                return {
                    userId: candidate.userId || candidate._id || candidate.id || null,
                    embedding
                };
            }

            return null;
        })
        .filter(Boolean);
};

// Sends a selfie plus the stored embeddings to the Python service, where
// the shop compares them and returns the closest match.
export const verifyFaceWithDeepFace = async (imageBuffer, storedEmbeddings) => {
    const normalizedCandidates = normalizeStoredCandidates(storedEmbeddings);

    if (!imageBuffer || normalizedCandidates.length === 0) {
        throw new Error("Selfie image and valid stored embeddings are required");
    }

    const form = new FormData();
    form.append("image", imageBuffer, { filename: "selfie.jpg" });
    form.append("stored_embeddings", JSON.stringify(normalizedCandidates));

    const { data } = await axios.post(`${PYTHON_SERVICE_URL}/verify`, form, {
        headers: form.getHeaders()
    });

    if (!data || typeof data.match !== "boolean") {
        throw new Error("Invalid response from face service");
    }

    return data;
};